### PR TITLE
Fix unsoundness on Windows

### PR DESCRIPTION
**Pull request by andrewhickman**
This PR fixes the unsound use of Cell and RefCell on windows (See #599). The first commit just replaces `Cell<bool>` with `AtomicBool` and moves `RefCell` into the `Mutex<WindowState>`. That should take care of the safety issues but I think there is still potential for the saved state to get out of sync with what has been executed. The second commit attempts to address this by moving all the window state variables into `WindowState` and moving all calls to `execute_in_thread` to within the lock of `WindowState`.

I think there are still some similar soundness issues on the macos backend ([e.g. here](https://github.com/tomaka/winit/blob/c1ef1acfc0a6d372519c3cf1a1c1e0696373fa61/src/platform/macos/window.rs#L529)) but I'm not currently able to test any fixes there.

### CODE REVIEW

**Comment by francesca64** [comment]
Sorry for the delay! This looks good, though I still haven't tested it out yet.

For soundness stuff like this, I think it's good to have a CHANGELOG entry, so people using older versions are aware of the unsoundness.

> moving all calls to `execute_in_thread` to within the lock of `WindowState`

If this is something we want to require from now on, then it would make sense for the parameters of `execute_in_thread` to enforce this.

**Comment by andrewhickman** [comment]
I've added a comment to `execute_in_thread`. There is one case where it is called outside of the lock (in Window::new) so I haven't fiddled with enforcing it further for now.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # Unreleased
+- Fixed some unsoundness around thread-safety on Windows.
```
Comment (francesca64) [inline CHANGELOG.md]: There should be a newline between the header and the list, and this entry is a bit vague.

File: src/platform/windows/events_loop.rs [hunk 2]
```diff
@@ -57,6 +57,7 @@ use platform::platform::dpi::{
     enable_non_client_dpi_scaling,
     get_hwnd_scale_factor,
 };
+use platform::platform::icon::WinIcon;
```
Comment (francesca64) [inline src/platform/windows/events_loop.rs]: Keeping with alphabetical order, this should go after `platform::platform::event` below.

File: src/platform/windows/window.rs [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
 
-use std::cell::{Cell, RefCell};
+use std::cell::Cell;
 use std::ffi::OsStr;
-use std::{io, mem, ptr};
 use std::os::windows::ffi::OsStrExt;
-use std::sync::{Arc, Mutex};
 use std::sync::mpsc::channel;
+use std::sync::{Arc, Mutex};
+use std::{io, mem, ptr};
```
Comment (francesca64) [inline src/platform/windows/window.rs]: I prefer to have the top-level brackety imports sorted above more specific ones... this also applies to the `winapi::um` below.

File: src/platform/windows/window.rs [hunk 4]
```diff
@@ -283,32 +272,23 @@ impl Window {
...
-            winuser::SetWindowLongW(
-                self.window.0,
-                winuser::GWL_STYLE,
-                style as _,
-            );
+            winuser::SetWindowLongW(self.window.0, winuser::GWL_STYLE, style as _);
         };
-        self.resizable.replace(resizable);
+    }
+        }
```
Comment (francesca64) [inline src/platform/windows/window.rs]: There's some very confusing bracket indentation happening here.

File: src/platform/windows/window.rs [hunk 5]
```diff
@@ -1029,6 +1003,13 @@ unsafe fn init(
             mouse_in_window: false,
             saved_window_info: None,
             dpi_factor,
+            fullscreen: attributes.fullscreen.clone(),
+            window_icon: window_icon,
```
Comment (francesca64) [inline src/platform/windows/window.rs]: Just `window_icon,` would be cleaner.

**Comment by francesca64** [comment]
I finally got around to testing this out, and things seem peachy. Just a few nitpicks, and this will make it into a release pretty soon.

**Comment by francesca64** [comment]
Thanks!

**Pull request merged**